source/slang/slang-ir-lower-generics.cpp

```diff
@@ -91,32 +107,160 @@ namespace Slang
             return loweredFunc;
         }
 
+        IRType* lowerGenericFuncType(IRBuilder* builder, IRGeneric* genericVal)
+        {
+            List<IRInst*> genericParamTypes;
+            for (auto genericParam : genericVal->getParams())
+            {
+                if (isPolymorphicType(genericParam->getFullType()))
+                {
+                    genericParamTypes.add(builder->getPtrType(builder->getVoidType()));
+                }
+                else
+                {
+                    genericParamTypes.add(genericParam->getFullType());
+                }
+            }
+
+            auto innerType = (IRFuncType*)lowerFuncType(
+                builder,
+                cast<IRFuncType>(findGenericReturnVal(genericVal)),
+                genericParamTypes.getCount());
+
+            for (int i = 0; i < genericParamTypes.getCount(); i++)
+            {
+                innerType->setOperand(
+                    innerType->getOperandCount() - genericParamTypes.getCount() + i,
+                    genericParamTypes[i]);
+            }
+
+            return innerType;
+        }
+
+        IRType* lowerFuncType(IRBuilder* builder, IRFuncType* funcType, UInt additionalParamCount = 0)
+        {
+            List<IRInst*> newOperands;
+            bool translated = false;
+            for (UInt i = 0; i < funcType->getOperandCount(); i++)
+            {
+                auto paramType = funcType->getOperand(i);
+                if (isPolymorphicType(paramType))
+                {
+                    newOperands.add(builder->getPtrType(builder->getVoidType()));
+                    translated = true;
+                }
+                else if (paramType->op == kIROp_Specialize)
+                {
+                    // TODO: handle static specialized type here.
+                    // For now treat all specialized types as dynamic.
+                    // In the future, we need to turn things like Array<IDynamic> into Array<void*>.
```

Comment:
As discussed briefly on another channel, there is a question of whether a parameter of type `Array<T>` should translate to an `Array<void*>` or to a single `void*`. There are arguments for and against each option.